File: python/linked_list.py
from typing import TypeVar, Generic, Optional, List

T = TypeVar('T')

class ListNode(Generic[T]):
    def __init__(self, value: T):
        self.value: T = value
        self.next: Optional[ListNode[T]] = None

class SinglyLinkedList(Generic[T]):
    def __init__(self):
        self._head: Optional[ListNode[T]] = None
        self._tail: Optional[ListNode[T]] = None
        self._length: int = 0

    @property
    def length(self) -> int:
        return self._length

    @property
    def is_empty(self) -> bool:
        return self._length == 0

    @property
    def head(self) -> Optional[T]:
        return self._head.value if self._head else None

    @property
    def tail(self) -> Optional[T]:
        return self._tail.value if self._tail else None

    def prepend(self, value: T) -> None:
        new_node = ListNode(value)
        if not self._head:
            self._head = new_node
            self._tail = new_node
        else:
            new_node.next = self._head
            self._head = new_node
        self._length += 1

    def append(self, value: T) -> None:
        new_node = ListNode(value)
        if not self._tail:
            self._head = new_node
            self._tail = new_node
        else:
            self._tail.next = new_node
            self._tail = new_node
        self._length += 1

    def remove_first(self) -> Optional[T]:
        if not self._head:
            return None
        removed_node = self._head
        self._head = self._head.next
        if not self._head:
            self._tail = None
        self._length -= 1
        return removed_node.value

    def clear(self) -> None:
        self._head = None
        self._tail = None
        self._length = 0

    def to_list(self) -> List[T]:
        result: List[T] = []
        current = self._head
        while current:
            result.append(current.value)
            current = current.next
        return result

    def __iter__(self):
        class Iterator:
            def __init__(self, head):
                self.current = head
            
            def __next__(self):
                if self.current is None:
                    raise StopIteration
                value = self.current.value
                self.current = self.current.next
                return value
        
        return Iterator(self._head)

    def __str__(self) -> str:
        return str(self.to_list())

class DoublyListNode(Generic[T]):
    def __init__(self, value: T):
        self.value: T = value
        self.next: Optional[DoublyListNode[T]] = None
        self.prev: Optional[DoublyListNode[T]] = None

class DoublyLinkedList(Generic[T]):
    def __init__(self):
        self._head: Optional[DoublyListNode[T]] = None
        self._tail: Optional[DoublyListNode[T]] = None
        self._length: int = 0

    @property
    def length(self) -> int:
        return self._length

    @property
    def is_empty(self) -> bool:
        return self._length == 0

    @property
    def head(self) -> Optional[T]:
        return self._head.value if self._head else None

    @property
    def tail(self) -> Optional[T]:
        return self._tail.value if self._tail else None

    def prepend(self, value: T) -> None:
        new_node = DoublyListNode(value)
        if not self._head:
            self._head = new_node
            self._tail = new_node
        else:
            new_node.next = self._head
            self._head.prev = new_node
            self._head = new_node
        self._length += 1

    def append(self, value: T) -> None:
        new_node = DoublyListNode(value)
        if not self._tail:
            self._head = new_node
            self._tail = new_node
        else:
            new_node.prev = self._tail
            self._tail.next = new_node
            self._tail = new_node
        self._length += 1

    def remove_first(self) -> Optional[T]:
        if not self._head:
            return None
        removed_node = self._head
        self._head = self._head.next
        if not self._head:
            self._tail = None
        else:
            self._head.prev = None
        self._length -= 1
        return removed_node.value

    def remove_last(self) -> Optional[T]:
        if not self._tail:
            return None
        removed_node = self._tail
        self._tail = self._tail.prev
        if not self._tail:
            self._head = None
        else:
            self._tail.next = None
        self._length -= 1
        return removed_node.value

    def clear(self) -> None:
        self._head = None
        self._tail = None
        self._length = 0

    def to_list(self) -> List[T]:
        result: List[T] = []
        current = self._head
        while current:
            result.append(current.value)
            current = current.next
        return result

    def to_list_reverse(self) -> List[T]:
        result: List[T] = []
        current = self._tail
        while current:
            result.append(current.value)
            current = current.prev
        return result

    def __iter__(self):
        class Iterator:
            def __init__(self, head):
                self.current = head
            
            def __next__(self):
                if self.current is None:
                    raise StopIteration
                value = self.current.value
                self.current = self.current.next
                return value
        
        return Iterator(self._head)

    def __str__(self) -> str:
        return str(self.to_list())

if __name__ == "__main__":
    print("Singly Linked List Example")
    sll = SinglyLinkedList[int]()
    sll.append(1)
    sll.append(2)
    sll.append(3)
    print(f"List: {sll}")
    print("Iterating with for loop:")
    for val in sll:
        print(val, end=' ')
    print()

    print("\nDoubly Linked List Example")
    dll = DoublyLinkedList[int]()
    dll.append(1)
    dll.append(2)
    dll.append(3)
    print(f"List: {dll}")
    print(f"Reverse: {dll.to_list_reverse()}")
    print("Iterating with for loop:")
    for val in dll:
        print(val, end=' ')
    print()

File: typescript/linked-list.ts
class ListNode<T> {
    value: T;
    next: ListNode<T> | null;

    constructor(value: T) {
        this.value = value;
        this.next = null;
    }
}

export class SinglyLinkedList<T> {
    private _head: ListNode<T> | null;
    private _tail: ListNode<T> | null;
    private _length: number;

    constructor() {
        this._head = null;
        this._tail = null;
        this._length = 0;
    }

    get length(): number {
        return this._length;
    }

    get isEmpty(): boolean {
        return this._length === 0;
    }

    get head(): T | undefined {
        return this._head?.value;
    }

    get tail(): T | undefined {
        return this._tail?.value;
    }

    prepend(value: T): void {
        const newNode = new ListNode(value);
        if (!this._head) {
            this._head = newNode;
            this._tail = newNode;
        } else {
            newNode.next = this._head;
            this._head = newNode;
        }
        this._length++;
    }

    append(value: T): void {
        const newNode = new ListNode(value);
        if (!this._tail) {
            this._head = newNode;
            this._tail = newNode;
        } else {
            this._tail.next = newNode;
            this._tail = newNode;
        }
        this._length++;
    }

    removeFirst(): T | undefined {
        if (!this._head) {
            return undefined;
        }
        const removedNode = this._head;
        this._head = this._head.next;
        if (!this._head) {
            this._tail = null;
        }
        this._length--;
        return removedNode.value;
    }

    clear(): void {
        this._head = null;
        this._tail = null;
        this._length = 0;
    }

    toArray(): T[] {
        const result: T[] = [];
        let current = this._head;
        while (current) {
            result.push(current.value);
            current = current.next;
        }
        return result;
    }

    [Symbol.iterator](): Iterator<T> {
        let current = this._head;
        return {
            next(): IteratorResult<T> {
                if (current === null) {
                    return { done: true, value: undefined };
                }
                const value = current.value;
                current = current.next;
                return { done: false, value };
            }
        };
    }
}

export class CircularLinkedList<T> {
    private _tail: ListNode<T> | null;
    private _length: number;

    constructor() {
        this._tail = null;
        this._length = 0;
    }

    get length(): number {
        return this._length;
    }

    get isEmpty(): boolean {
        return this._length === 0;
    }

    get head(): T | undefined {
        return this._tail?.next?.value;
    }

    get tail(): T | undefined {
        return this._tail?.value;
    }

    prepend(value: T): void {
        const newNode = new ListNode(value);
        if (!this._tail) {
            newNode.next = newNode;
            this._tail = newNode;
        } else {
            newNode.next = this._tail.next;
            this._tail.next = newNode;
        }
        this._length++;
    }

    append(value: T): void {
        const newNode = new ListNode(value);
        if (!this._tail) {
            newNode.next = newNode;
            this._tail = newNode;
        } else {
            newNode.next = this._tail.next;
            this._tail.next = newNode;
            this._tail = newNode;
        }
        this._length++;
    }

    removeFirst(): T | undefined {
        if (!this._tail) {
            return undefined;
        }
        const removedNode = this._tail.next!;
        if (this._tail === removedNode) {
            this._tail = null;
        } else {
            this._tail.next = removedNode.next;
        }
        this._length--;
        return removedNode.value;
    }

    clear(): void {
        this._tail = null;
        this._length = 0;
    }

    toArray(): T[] {
        const result: T[] = [];
        if (!this._tail) {
            return result;
        }
        let current = this._tail.next;
        for (let i = 0; i < this._length; i++) {
            if (current) {
                result.push(current.value);
                current = current.next;
            }
        }
        return result;
    }

    [Symbol.iterator](): Iterator<T> {
        let current = this._tail?.next || null;
        let count = 0;
        const length = this._length;
        return {
            next(): IteratorResult<T> {
                if (count >= length || current === null) {
                    return { done: true, value: undefined };
                }
                const value = current.value;
                current = current.next;
                count++;
                return { done: false, value };
            }
        };
    }
}

class DoublyListNode<T> {
    value: T;
    next: DoublyListNode<T> | null;
    prev: DoublyListNode<T> | null;

    constructor(value: T) {
        this.value = value;
        this.next = null;
        this.prev = null;
    }
}

export class DoublyLinkedList<T> {
    private _head: DoublyListNode<T> | null;
    private _tail: DoublyListNode<T> | null;
    private _length: number;

    constructor() {
        this._head = null;
        this._tail = null;
        this._length = 0;
    }

    get length(): number {
        return this._length;
    }

    get isEmpty(): boolean {
        return this._length === 0;
    }

    get head(): T | undefined {
        return this._head?.value;
    }

    get tail(): T | undefined {
        return this._tail?.value;
    }

    prepend(value: T): void {
        const newNode = new DoublyListNode(value);
        if (!this._head) {
            this._head = newNode;
            this._tail = newNode;
        } else {
            newNode.next = this._head;
            this._head.prev = newNode;
            this._head = newNode;
        }
        this._length++;
    }

    append(value: T): void {
        const newNode = new DoublyListNode(value);
        if (!this._tail) {
            this._head = newNode;
            this._tail = newNode;
        } else {
            newNode.prev = this._tail;
            this._tail.next = newNode;
            this._tail = newNode;
        }
        this._length++;
    }

    removeFirst(): T | undefined {
        if (!this._head) {
            return undefined;
        }
        const removedNode = this._head;
        this._head = this._head.next;
        if (this._head) {
            this._head.prev = null;
        } else {
            this._tail = null;
        }
        this._length--;
        return removedNode.value;
    }

    removeLast(): T | undefined {
        if (!this._tail) {
            return undefined;
        }
        const removedNode = this._tail;
        this._tail = this._tail.prev;
        if (this._tail) {
            this._tail.next = null;
        } else {
            this._head = null;
        }
        this._length--;
        return removedNode.value;
    }

    clear(): void {
        this._head = null;
        this._tail = null;
        this._length = 0;
    }

    toArray(): T[] {
        const result: T[] = [];
        let current = this._head;
        while (current) {
            result.push(current.value);
            current = current.next;
        }
        return result;
    }

    toArrayReverse(): T[] {
        const result: T[] = [];
        let current = this._tail;
        while (current) {
            result.push(current.value);
            current = current.prev;
        }
        return result;
    }

    [Symbol.iterator](): Iterator<T> {
        let current = this._head;
        return {
            next(): IteratorResult<T> {
                if (current === null) {
                    return { done: true, value: undefined };
                }
                const value = current.value;
                current = current.next;
                return { done: false, value };
            }
        };
    }
}

export class DoublyCircularLinkedList<T> {
    private _tail: DoublyListNode<T> | null;
    private _length: number;

    constructor() {
        this._tail = null;
        this._length = 0;
    }

    get length(): number {
        return this._length;
    }

    get isEmpty(): boolean {
        return this._length === 0;
    }

    get head(): T | undefined {
        return this._tail?.next?.value;
    }

    get tail(): T | undefined {
        return this._tail?.value;
    }

    prepend(value: T): void {
        const newNode = new DoublyListNode(value);
        if (!this._tail) {
            newNode.next = newNode;
            newNode.prev = newNode;
            this._tail = newNode;
        } else {
            newNode.next = this._tail.next;
            newNode.prev = this._tail;
            this._tail.next!.prev = newNode;
            this._tail.next = newNode;
        }
        this._length++;
    }

    append(value: T): void {
        const newNode = new DoublyListNode(value);
        if (!this._tail) {
            newNode.next = newNode;
            newNode.prev = newNode;
            this._tail = newNode;
        } else {
            newNode.next = this._tail.next;
            newNode.prev = this._tail;
            this._tail.next!.prev = newNode;
            this._tail.next = newNode;
            this._tail = newNode;
        }
        this._length++;
    }

    removeFirst(): T | undefined {
        if (!this._tail) {
            return undefined;
        }
        const removedNode = this._tail.next!;
        if (this._tail === removedNode) {
            this._tail = null;
        } else {
            this._tail.next = removedNode.next;
            removedNode.next!.prev = this._tail;
        }
        this._length--;
        return removedNode.value;
    }

    removeLast(): T | undefined {
        if (!this._tail) {
            return undefined;
        }
        const removedNode = this._tail;
        if (this._tail === removedNode.next) {
            this._tail = null;
        } else {
            this._tail = removedNode.prev!;
            this._tail.next = removedNode.next;
            removedNode.next!.prev = this._tail;
        }
        this._length--;
        return removedNode.value;
    }

    clear(): void {
        this._tail = null;
        this._length = 0;
    }

    toArray(): T[] {
        const result: T[] = [];
        if (!this._tail) {
            return result;
        }
        let current = this._tail.next;
        for (let i = 0; i < this._length; i++) {
            if (current) {
                result.push(current.value);
                current = current.next;
            }
        }
        return result;
    }

    toArrayReverse(): T[] {
        const result: T[] = [];
        if (!this._tail) {
            return result;
        }
        let current: DoublyListNode<T> | null = this._tail;
        for (let i = 0; i < this._length; i++) {
            if (current) {
                result.push(current.value);
                current = current.prev;
            }
        }
        return result;
    }

    [Symbol.iterator](): Iterator<T> {
        let current = this._tail?.next || null;
        let count = 0;
        const length = this._length;
        return {
            next(): IteratorResult<T> {
                if (count >= length || current === null) {
                    return { done: true, value: undefined };
                }
                const value = current.value;
                current = current.next;
                count++;
                return { done: false, value };
            }
        };
    }
}

console.log("Singly Linked List Example");
const sll = new SinglyLinkedList<number>();
sll.append(1);
sll.append(2);
sll.append(3);
console.log(`List: [${sll.toArray().join(", ")}]`);
console.log("Iterating with for loop:");
let output = '';
for (const val of sll) {
    output += val + ' ';
}
console.log(output.trim());

console.log("\nCircular Linked List Example");
const cll = new CircularLinkedList<number>();
cll.append(1);
cll.append(2);
cll.append(3);
console.log(`List: [${cll.toArray().join(", ")}]`);
console.log("Iterating with for loop:");
output = '';
for (const val of cll) {
    output += val + ' ';
}
console.log(output.trim());

console.log("\nDoubly Linked List Example");
const dll = new DoublyLinkedList<number>();
dll.append(1);
dll.append(2);
dll.append(3);
console.log(`List: [${dll.toArray().join(", ")}]`);
console.log(`Reverse: [${dll.toArrayReverse().join(", ")}]`);
console.log("Iterating with for loop:");
output = '';
for (const val of dll) {
    output += val + ' ';
}
console.log(output.trim());

console.log("\nDoubly Circular Linked List Example");
const dcll = new DoublyCircularLinkedList<number>();
dcll.append(1);
dcll.append(2);
dcll.append(3);
console.log(`List: [${dcll.toArray().join(", ")}]`);
console.log(`Reverse: [${dcll.toArrayReverse().join(", ")}]`);
console.log("Iterating with for loop:");
output = '';
for (const val of dcll) {
    output += val + ' ';
}
console.log(output.trim());
